Add memberlist right-panel to Hydrogen

MidhunSureshR opened the pull request.

MidhunSureshR committed:
Basic barebones of memberlist view

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Duplicate navigation from room details

- TODO: Remove duplication

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move rightpanel to separate view and vm

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Modify navigation to work with rightpanel segment

- Ensure that rightpanel does not show in the URL.
- Create an action of details to insert the rightpanel segment.
- Make sure rightpanel can be a child of room.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Create RightPanel from SessionViewModel

- Also remove old methods used in RoomDetails

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Accommodate "rightpanel" navigation segment

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Modify CSS to reflect changes

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add memberlist to navigation

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove code from session view/vm

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Create memberlist view from rightpanel

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove unused param

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make list scrollable

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make member private

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make getUserLevel() public

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Getter for powerlevel from room

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Consider powerlevels in comparator and add tests

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Sort memberlist using powerlevel

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use name instead of displayName

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Implement set method to support update from value

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use set instead of add in memberlist

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add binding in view

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add update method

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use MappedList instead of MappedMap

- MappedList supports an optional updater method

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Support updater in MappedMap

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Settle on MappedMap

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix bug in FilteredMap

Caused multiple updates to be triggered.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Implement name disambiguation

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make tile vm support disambiguation

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Call disambiguate on update/map

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Put up a temporary loading view

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Switch to collator for perf reasons

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add avatar to tile

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add some initial styling

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move loading view into separate file

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Support loading view for all panels

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove duplication in RightPanelViewModel

MidhunSureshR committed:
Disambiguator is not async

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
rename rightpanel to right-panel

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove duplication

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add explaining comment

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
name changes

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/domain/session/leftpanel/LeftPanelViewModel.js:
```diff
@@ -93,8 +93,13 @@ export class LeftPanelViewModel extends ViewModel {
+        let _path = path;
         const details = this.navigation.path.get("details");
-        return details?.value ? path.with(details) : path;
+        if (details?.value) {
```
This will need to be generic to copy any segments over that come after a `right-panel` segment.

MidhunSureshR replied:
Yep. Haven't gotten around to making the navigation work everywhere. Planning on doing it right after lazy loading. 👍

bwindels commented on src/domain/session/rightpanel/comparator.js:
```diff
@@ -0,0 +1,98 @@
+import {PowerLevels} from "../../../matrix/room/timeline/PowerLevels.js";
```
Perhaps put this file (comparator.js) in a `members` subdirectory?

bwindels commented on src/domain/session/rightpanel/disambiguator.js:
```diff
@@ -0,0 +1,147 @@
+export class Disambiguator {
```
same here, can be in `members` sub dir

bwindels commented on src/domain/session/rightpanel/disambiguator.js:
```diff
@@ -0,0 +1,147 @@
+
+    _updateMap(vm) {
+        const name = vm.name;
+        if (this._map.has(name)) {
```
You can prevent the double lookup here by doing `get` immediately and check if the result is truthy

bwindels commented on src/domain/session/rightpanel/disambiguator.js:
```diff
@@ -0,0 +1,147 @@
+    disambiguate(vm) {
+        if (!vm.nameChanged) { return; }
+        this._handlePreviousName(vm);
+        this._updateMap(vm);
```
You could return the array of `vm`s from this method (only if it is an array), so you don't need to look it up again here.

bwindels commented on src/domain/session/rightpanel/disambiguator.js:
```diff
@@ -0,0 +1,147 @@
+
+    _handlePreviousName(vm) {
+        const previousName = vm.previousName;
+        if (!previousName) { return; }
```
Perhaps check that previousName is not a string or is not undefined rather than falsy in case somebody has an empty name

bwindels commented on src/domain/session/rightpanel/disambiguator.js:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    _flatten(name, array) {
+        const vm = array.pop();
```
Can we write this inline in `_handlePreviousName` perhaps? I find it a bit harder to read when this is disconnected from the `length === 1` check. Then you could do `array[0]` for more clarity.

bwindels commented on src/platform/web/ui/session/rightpanel/LoadingView.js:
```diff
@@ -0,0 +1,8 @@
+import {TemplateView} from "../../general/TemplateView.js";
+import {spinner} from "../../common.js";
+
+export class LoadingView extends TemplateView {
```
I wonder if it is worth it to create a generic loading view that accepts a message and perhaps an extra css class, and by default just center the message and spinner on both axis, like `TimelineLoadingView`.

MidhunSureshR replied:
This is pretty much what I thought too. We'd then replace TimelineLoadingView with the generic loading view.

bwindels replied:
Ok, but we're not doing it in this PR I suppose? fine be me

bwindels commented on src/platform/web/ui/session/rightpanel/MemberListView.js:
```diff
@@ -0,0 +1,9 @@
+
+export class MemberListView extends TemplateView {
+    render(t, vm) {
+        return t.view(new ListView({list: vm.memberTileViewModels, className:"MemberListView"}, tileViewModel => new MemberTileView(tileViewModel)));
```
You can just define MemberListView as a sub class of `ListView`, providing a convenience constructor. Or do you plan to add more things to this view?

MidhunSureshR replied:
Will keep this in mind. But some of this is bound to change with lazy loading I think.

MidhunSureshR replied:
Nothing changed with lazy loading :sweat_smile:

bwindels commented on src/matrix/room/BaseRoom.js:
```diff
@@ -378,6 +378,10 @@ export class BaseRoom extends EventEmitter {
     }
 
+    get powerLevels() {
+        return this._timeline.powerLevels;
```
`this._timeline` is only set if the timeline is open, which will be fine for using it in the in the member list, but it will throw if `openTimeline` is not called before `loadMemberlist`. Ideally we'd:

 - lazy load the power levels in BaseRoom either when loading the members or opening the timeline, and pass them into Timeline and Memberlist.
 - if loaded, update the power levels in afterSync of Room
 - resort the member list when the power levels change in afterSync.

bwindels reviewed (commented):
This is looking great! I tried it out and the sorting and disambiguation seemed to work quite well.
I came across a few things in this first review, see the comments.

MidhunSureshR committed:
Eliminate double lookup

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Return array to prevent fetching again

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add failing test

For empty names not un-disambiguating

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Check prev name is string

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Inline flatten method

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR commented on src/domain/session/rightpanel/MemberListViewModel.js:
```diff
@@ -0,0 +1,35 @@
+        }
+        const updater = (vm, params, newMember) => {
+            vm.updateFrom(newMember);
+            this.nameDisambiguator.disambiguate(vm);
```
Noticed something here. `MappedMap` won't run the mapper in `onRemove`. This will cause any user who leaves and rejoins a room to appear disambiguated.

bwindels replied:
Ah yes, can you check for this in the mapper when it is run in onAdd?

MidhunSureshR replied:
Fixed for now but there are related issues. For instance if user-A and user-B are disambiguated and user-B leaves the room, we'd ideally want user-A to be un-disambiguated (which we can't really do atm because the disambiguator does not get informed when user-B leaves).

bwindels replied:
good point, I think this is fine though for now.

MidhunSureshR committed:
Support slice

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Export function

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add LazyListView

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Switch to lazy list

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,128 @@
+        if (!this._list) {
+            return;
+        }
+        this._subscription = this._list.subscribe(this);
```
any reason you can't leave this up to the parent class by calling `super`? Should be enough to just override `onAdd`, ...?

MidhunSureshR replied:
Because `super.loadList()` also inserts all the elements from list into `_root`.

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,128 @@
+    constructor({itemHeight, height, appendCount = 5, ...options}, childCreator) {
+        super(options, childCreator);
+        this._itemHeight = itemHeight;
+        this._height = height;
```
I'd try to determine this inside this component using the DOM rather than inject (& hardcode) it. It does mean you might need to make the initial rendering async as in render at first you'd only attach the `ul` to measure the height of it, and only then mount the needed amount of children. In my experience, doing a double `await new Promise(r => requestAnimationFrame(r))` does a good job of reading a reliably `clientHeight`.

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,128 @@
+        if (!this._isFullyScrolled()) {
+            return;
+        }
+        this._range.expandFromEnd(this._appendCount);
```
I think this will add `n` (default of 5) items *every* time you scroll a pixel? I'd like to interpret this number as a padding rather, where you have the range the corresponds to the current viewport and you add `n` items on either end (limited by 0 and the length of the list on either end of course).

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    _onInitialExpand() {
+        if (this._expanded) { return; }
```
What does expand mean here? Move the window of the range to react to scrolling? Is it inspired [on this](https://github.com/matrix-org/matrix-react-sdk/blob/9d6d8fc666855ca0c06c71ccb30c74ac4fc8fd12/src/components/views/elements/LazyRenderList.js#L39)? It seems to do something somewhat different though.

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,128 @@
+            this.onBeforeListChanged();
+            const [child] = this._childInstances.splice(fromIdx, 1);
+            child.root().remove();
+            this.onListChanged();
```
you'll need to render an item at the end of the list though to keep the count the same

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,128 @@
+            this.onBeforeListChanged();
+            const child = this._childCreator(value);
+            this._childInstances.splice(toIdx, 0, child);
+            insertAt(this._root, toIdx, mountView(child, this._mountArgs));
```
same here, you'll need to remove an item at the end of the list to keep the count the same

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    expandFromEnd(units) {
+        this._onInitialExpand();
```
I'm failing to understand what the idea here is

bwindels reviewed (commented):
Nice work so far, left some comments on the lazy list view.

MidhunSureshR committed:
Incorporate lazyrender code from element

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix updates

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add explaining comment

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Improve comment

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove misleading comment

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix update method

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use normalized index in recreateItem

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Find height of container from DOM

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Keep memberlist panel open on room/grid change

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/session/rightpanel/MemberTileView.js:
```diff
@@ -0,0 +1,11 @@
+
+export class MemberTileView extends TemplateView {
+    render(t, vm) {
+        return t.div({ className: "MemberTileView" }, [
```
the root element should probably be an `<li>` here as it's mounted within an `<ul>`.

MidhunSureshR committed:
div --> li

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Account for padding in itemHeight

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Replace slice with iterator

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,250 @@
+        const range = this._getVisibleRange();
+        const intersectRange = range.expand(this._overflowMargin);
+        const renderRange = range.expand(this._overflowItems);
+        const listHasChangedSize = !!this._renderRange && this._list.length !== this._renderRange.totalSize();
```
`listHasChangedSize` can be removed here and in the check below, as any changes to the list will be handled in the onAdd/onMove/onRemove/... handlers.

MidhunSureshR replied:
`listHasChangedSize` is removed but we still need an option to force render for the time being until we optimize onAdd/onRemove.

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,250 @@
+    }
+}
+
+export class LazyListView extends ListView {
```
I wonder how much behaviour is still used from `ListView`? Do you think it could be simpler to just implement this from scratch?

MidhunSureshR replied:
Very little. It would me much simpler to implement this from scratch indeed. You can see in some places (like `loadList`) where the code gets a little crufty because we have to support the super class.

MidhunSureshR committed:
Make powerLevels observable

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/matrix/room/BaseRoom.js:
```diff
@@ -378,8 +381,44 @@ export class BaseRoom extends EventEmitter {
 
-    get powerLevels() {
-        return this._timeline.powerLevels;
+    async loadPowerLevels() {
```
this should be private (e.g. _load...)

bwindels commented on src/matrix/room/BaseRoom.js:
```diff
@@ -378,8 +381,44 @@ export class BaseRoom extends EventEmitter {
-    get powerLevels() {
-        return this._timeline.powerLevels;
+    async loadPowerLevels() {
+        const timelineReader = new TimelineReader({
```
The timeline reader is not needed for the power levels. This is the part the reads the messages in the Timeline class.

bwindels commented on src/matrix/room/BaseRoom.js:
```diff
@@ -378,8 +381,44 @@ export class BaseRoom extends EventEmitter {
+    }
+
+    async observePowerLevels() {
+        let observable = this._powerLevels;
```
as the function currently is, if a second call is made before `load...` finishes for the first call, it will load the power levels twice. I think you will need to cache the promise from `load...` and store it on `this` and await that if it is set.

bwindels commented on src/matrix/room/BaseRoom.js:
```diff
@@ -378,8 +381,44 @@ export class BaseRoom extends EventEmitter {
+            fragmentIdComparer: this._fragmentIdComparer
+        });
+        const txn = await this._storage.readTxn(
+            timelineReader.readTxnStores.concat(this._storage.storeNames.roomMembers, this._storage.storeNames.roomState)
```
the transaction should only use the `roomState` store, as that is the only store you'll be reading from here.

MidhunSureshR committed:
Remove timeline reader and only use roomState

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make loadPowerLevels private

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Ensure that power levels are loaded only once

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix broken tests

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Unmount child views correctly

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove listHasChangedSize

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move css to layout.css

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Reduce padding

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/matrix/room/BaseRoom.js:
```diff
@@ -404,11 +404,14 @@ export class BaseRoom extends EventEmitter {
     }
 
     async observePowerLevels() {
+        if (this._powerLevelLoading) { await this._powerLevelLoading; }
```
new members should be set to null in the constructor, I think this could cause v8 to de-optimize our code if the hidden class doesn't match any more.

MidhunSureshR committed:
Create UI to open memberlist from details panel

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add chevron to button

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Implement UX

- Add chevron to member count in room details
- Make some rows in panel buttons
- Add user chrome to right panel
- Style UI

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix lazy list css

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make addPanelIfNeeded more generic

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Center names

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/domain/session/leftpanel/LeftPanelViewModel.js:
```diff
@@ -92,9 +93,10 @@ export class LeftPanelViewModel extends ViewModel {
-    _pathForDetails(path) {
-        const details = this.navigation.path.get("details");
-        return details?.value ? path.with(details) : path;
+    _pathForRightPanel(path) {
```
why not just inline `path = addPanelIfNeeded(this.navigation, path);` where this is used?

MidhunSureshR committed:
Style loading view

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Release memberlist after panel is closed

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add jsdoc for powerlevels

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove comment

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Subscribe to powerLevels

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move spinner before the text

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/session/rightpanel/RightPanelView.js:
```diff
@@ -0,0 +1,34 @@
+        };
+        return t.div({ className: "RightPanelView" },
+            [
+                t.mapView(vm => vm.activeViewModel && vm, vm => vm ? new ButtonsView(vm) : null),
```
you could use `ifView` here

MidhunSureshR committed:
Css fixes

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Clear margin on ul

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
More css fixes

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Remove top padding of room details view

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/css/layout.css:
```diff
@@ -208,3 +208,13 @@ the layout viewport up without resizing it when the keyboard shows */
 }
+
+.LazyListParent {
+    height: 100vh;
```
Not sure why this doesn't make the lazy list too tall, but 100vh doesn't seem correct as there are buttons above it. This would be better:
```suggestion
    flex: 1;
```

bwindels commented on src/platform/web/ui/css/right-panel.css:
```diff
@@ -1,6 +1,13 @@
+
+.RoomDetailsView {
     flex-direction: column;
+    height: 100%;
```
same here
```suggestion
    flex: 1;
```

bwindels commented on src/domain/session/rightpanel/MemberListViewModel.js:
```diff
@@ -0,0 +1,46 @@
+
+    get previousSegmentName() { return "details"; }
+
+    _filterJoinedMembers(members) {
```
not sure it's worth to put this in a method? seems easier to read if the whole chain is in one place ...

MidhunSureshR committed:
Remove TemplateView

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move files to members directory

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move LoadingView.js

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/domain/session/rightpanel/RightPanelViewModel.js:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    _setupUpdater(segment, viewmodel, argCreator) {
+        const updater = async (skipDispose = false) => {
```
why do we need this flag? wouldn't only one of the options have `enable` true below?

MidhunSureshR replied:
We have to run the `updater` for both the segments ("memberlist" and "details") once initially because otherwise by the time the `RightPanelViewModel` is created and the segment is observed, the navigation event would have already fired.
Only one of the initial `updater` runs would actually create the `activeViewModel` but we don't want the second initial updater call to dispose of the view model created by an earlier initial updater call.

bwindels replied:
right, makes sense now :+1:

bwindels commented on src/domain/session/rightpanel/comparator.js:
```diff
@@ -0,0 +1,97 @@
+
+export function createMemberComparator(powerLevels) {
+    const collator = new Intl.Collator();
+    const removeCharacter = string => string.charAt(0) === "@"? string.slice(1) : string;
```
singling out `@` seems a bit arbitrary here, but probably fine for now.

MidhunSureshR replied:
Perhaps removing the explaining comment from here just hours ago was a mistake :smile:
This is so that we can spread the members whose name is just their user id (starting with @) from clubbing together at the top of the list. By ignoring the starting @, these members are alphabetically sorted on their user id.

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,251 @@
+        await new Promise(r => requestAnimationFrame(r));
+
+        this._height = this._parent.clientHeight;
+        if(this._height === 0) { throw "LazyListView could not calculate parent height."}
```
this method should not throw as it is not awaited from `mount`, better just log?

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,251 @@
+        this._renderElementsInRange();
+    }
+
+    _renderItems(items) {
```
since this method being called from anywhere but _renderElementsInRange would break things (as _renderElementsInRange clears the DOM), should we merge the two?

MidhunSureshR replied:
Have merged the two for the time being. The follow up PR will change the render code anyway.

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,251 @@
+            // Element is moved up the list, so render element from top boundary
+            const index = topCount;
+            const child = childFromIndex(index);
+            // Modify childInstances here
```
should we do this in this PR? Sounds like it would create bugs otherwise?

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,251 @@
+
+    _renderAdditionalElement(fromIdx, toIdx) {
+        const {topCount, renderCount} = this._renderRange;
+        const childFromIndex = index => this._childCreator(this._list.get(index));
```
`BaseObservableList` doesn't define a `get` method ... should we also use an iterator here for now?

bwindels replied:
(also, a lambda for this seems overkill?)

MidhunSureshR replied:
> `BaseObservableList` doesn't define a `get` method ... should we also use an iterator here for now?

Eventually we'd want a `get` method for `BaseObservableList` no? We sure can do iterator for now.

bwindels replied:
Yeah, probably. Just need to think if it can be implemented well in all observable list implementations (e.g. filtering changes the index).

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,251 @@
+        const childFromIndex = index => this._childCreator(this._list.get(index));
+        if (toIdx < fromIdx) {
+            // Element is moved up the list, so render element from top boundary
+            const index = topCount;
```
I don't understand why toIdx and fromIdx are seemingly being ignored here. If an element moves within the current rendered range, I'd think we'd need to take the specific indices into account?

bwindels replied:
Oooh, I see. This is about adding an item at either edge after removing one to keep the count in the rendered range the same. Perhaps we can make this clearer?

bwindels commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,251 @@
+        }
+    }
+
+    _renderAdditionalElement(fromIdx, toIdx) {
```
Can we give this a name more specific to moving an item? Given it's arguments, it looks like it would only apply in that scenario.

bwindels reviewed (commented):
looking good, we're almost there!

MidhunSureshR committed:
Inline method

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use ifView instead of mapView

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use flex instead of setting height

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Inline method

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Log instead of throwing error

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Inline method

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR commented on src/platform/web/ui/general/LazyListView.js:
```diff
@@ -0,0 +1,251 @@
+    _removeAdditionalElement(fromIdx, toIdx) {
+        if (toIdx < fromIdx) {
+            // Element comes from the bottom, so remove element at bottom
+            this._root.lastChild.remove();
```
Unmount child and modify `childInstances` here.

MidhunSureshR committed:
Fix lazylist

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Improve code

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix move bug

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Replace get with iterator and remove lambda

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make code clearer

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix disambiguator

- Do not disambiguate name on room rejoin

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Initialize prop in constructor

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix layout of details panel in mobile

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix width for smaller screens

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Import from AvatarView.js

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>
